sklearn/decomposition/dict_learning.py

```diff
@@ -110,7 +110,8 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
         clf = Lasso(alpha=alpha, fit_intercept=False, precompute=gram,
                     max_iter=max_iter, warm_start=True)
         clf.coef_ = init
-        clf.fit(dictionary.T, X.T)
+        # Copying X to avoid buffer source-array read only error
+        clf.fit(dictionary.T, X.copy().T)
```

Comment:
Debugging this with the data that causes error, I cannot see any indication on X being read only. 
```python
X.flags['writeable'] = True
```
Adding a unittest that would currently fail could be tricky as the error is only triggered by very large X.